networking/gcp/vpc/main.tf

```
@@ -0,0 +1,121 @@
+  description              = "Public subnet of IP address range ${each.key} in VPC ${google_compute_network.vpc.name}"
+  purpose                  = "PRIVATE_RFC_1918"
+  private_ip_google_access = var.enable_google_access
+  region                   = data.google_client_config.current.region
+  project                  = data.google_client_config.current.project
+  log_config {
+    aggregation_interval = var.flow_log_max_aggregation_interval
+    flow_sampling        = "1.0"
+    metadata             = "INCLUDE_ALL_METADATA"
+  }
+}
+
+# GKE private subnets
+resource "google_compute_subnetwork" "gke_subnets" {
+  for_each                 = var.gke_subnets
+  ip_cidr_range            = each.value.nodes_cidr_block
+  name                     = each.key
+  network                  = google_compute_network.vpc.name
+  description              = "GKE private subnet of IP address range ${each.value.nodes_cidr_block} in VPC ${google_compute_network.vpc.name}"
+  purpose                  = "PRIVATE_RFC_1918"
+  private_ip_google_access = var.enable_google_access
+  region                   = data.google_client_config.current.region
+  project                  = data.google_client_config.current.project
+  log_config {
+    aggregation_interval = var.flow_log_max_aggregation_interval
+    flow_sampling        = "1.0"
+    metadata             = "INCLUDE_ALL_METADATA"
+  }
+  dynamic "secondary_ip_range" {
+    for_each = ["pods_cidr_block", "services_cidr_block"]
+    content {
+      ip_cidr_range = each.value[secondary_ip_range.key]
+      range_name    = secondary_ip_range.key == "pods_cidr_block" ? "${each.key}-pod-ranges" : "${each.key}-svc-ranges"
+    }
+  }
+}
+
+# External access for public subnets
+# Router
+resource "google_compute_router" "router" {
+  count       = local.enable_external_access ? 1 : 0
+  name        = "${google_compute_network.vpc.name}-router"
+  network     = google_compute_network.vpc.self_link
+  description = "Router in the VPC ${google_compute_network.vpc.name}"
+  region      = data.google_client_config.current.region
+  project     = data.google_client_config.current.project
+}
+
+# NAT gateway
+resource "google_compute_router_nat" "nat_gateway" {
+  for_each                           = length(google_compute_router.router) > 0 ? [1] : []
+  name                               = "${google_compute_network.vpc.name}-nat"
+  nat_ip_allocate_option             = "AUTO_ONLY"
+  source_subnetwork_ip_ranges_to_nat = "LIST_OF_SUBNETWORKS"
+  router                             = one(google_compute_router.router[*].name)
+  region                             = data.google_client_config.current.region
+  project                            = data.google_client_config.current.project
+  dynamic "subnetwork" {
+    for_each = google_compute_subnetwork.public_subnets
+    content {
+      name                    = subnetwork.value.name
+      source_ip_ranges_to_nat = ["PRIMARY_IP_RANGE"]
+    }
+  }
```

If you merge the definitions of the public and private subnets, you would need to do the following here:
```suggestion
  dynamic "subnetwork" {
    for_each = var.public_subnets
    content {
      name                    = google_compute_subnetwork.subnets[subnetwork.value].name
      source_ip_ranges_to_nat = ["PRIMARY_IP_RANGE"]
    }
  }
```

---

done